src/main/java/hudson/remoting/AsyncFutureImpl.java

```diff
@@ -78,13 +79,28 @@ public synchronized V get() throws InterruptedException, ExecutionException {
         return value;
     }
 
+    @CheckForNull
     public synchronized V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        if(!completed)
-            wait(unit.toMillis(timeout));
+        // The accuracy of wait(long) operation is milliseconds anyway, but ok.
+        long endWaitTime = System.nanoTime() + unit.toNanos(timeout);
+        while (!completed) {
+            long timeToWait = endWaitTime - System.nanoTime();
+            if (timeout < 0) {
+                break;
+            }
+
+            long timeToWaitMs = Math.max(1, timeToWait / 1000000);
+            wait(timeToWaitMs);
```

Comment:
There is `wait(long,int)` to wait with “nanosecond” precision

Reply:
The implementation in my Java does the same, actually. Decompiled code:

```
public final void wait(long timeout, int nanos) throws InterruptedException {
        if (timeout < 0) {
            throw new IllegalArgumentException("timeout value is negative");
        }

        if (nanos < 0 || nanos > 999999) {
            throw new IllegalArgumentException(
                                "nanosecond timeout value out of range");
        }

        if (nanos >= 500000 || (nanos != 0 && timeout == 0)) {
            timeout++;
        }

        wait(timeout);
    }
```

I will patch the impl though